model/messages/collection.go

```diff
@@ -24,12 +24,70 @@ type CollectionResponse struct {
 	Nonce      uint64 // so that we aren't deduplicated by the network layer
 }
 
+// UntrustedClusterBlockPayload is a duplicate of cluster.Payload used within
+// untrusted messages. It exists only to provide a memory-safe structure for
+// decoding messages and should be replaced in the future by updating the core
+// cluster.Payload type.
+// Deprecated: Please update cluster.Payload.Collection to use []flow.TransactionBody,
+// then replace instances of this type with cluster.Payload
+type UntrustedClusterBlockPayload struct {
+	Collection       []flow.TransactionBody
+	ReferenceBlockID flow.Identifier
+}
+
+// UntrustedClusterBlock is a duplicate of cluster.Block used within
+// untrusted messages. It exists only to provide a memory-safe structure for
+// decoding messages and should be replaced in the future by updating the core
+// cluster.Block type.
+// Deprecated: Please update cluster.Payload.Collection to use []flow.TransactionBody,
+// then replace instances of this type with cluster.Block
+type UntrustedClusterBlock struct {
```

Comment:
Interested if we can use generics to deal with cluster blocks? Can we give it a try?

Reply:
I did try this initially but didn't really find a simple way to apply generics here. It might be easier if `flow.Block` used generics too...

It isn't compiling but I'll throw it in `jordan/3374--generics-experimentation-mess` if you want to take a look at what I was trying. Maybe you have some different ideas for how to approach it.